database/messages.js

```diff
@@ -1,69 +1,80 @@
 // connect to database
-require('./db');
+var db = require('./db');
 
 var mongoose = require('mongoose');
 var Promise = require('bluebird');
 mongoose.Promise = Promise;
 var messages = {};
 
 var MessageSchema = new mongoose.Schema({
+	id: Number,
 	text: String,
   	user: String,
   	reported: {type: Boolean, default: false},
   	date: { type: Date, default: Date.now }
 });
+MessageSchema.index({ id: 1});
 
 var MessageModel = mongoose.model('messages', MessageSchema);
 
-messages.saveMessage = function(data) {
-	var message = new MessageModel(
-	{
-		text: data.text,
-		user: data.user
+messages.saveMessage = function(data, callback) {
+	db.getNextAvailableId(db.messageIdKey, function(nextId) {
+		// create a new message object with given data and a unique id
+		var message = new MessageModel({
+			id: nextId,
+			text: data.text,
+			user: data.user
+		})
+		message.save();
+
+		db.setLastTakenId(db.messageIdKey, nextId, function(err) {
+			if (err) { console.log("error setting next id for messages: ", err); }
+        });
+
+        callback(message);
 	})
-	message.save();
 };
 
 messages.getReportedMessages = function(callback) {
-	var param = {reported: true};
-	MessageModel.find(param, function(err, data) {
-		console.log(data);
+	MessageModel.find({reported: true}, function(err, data) {
 		callback(data);
 	});
 }
 
-messages.report = function(user, text, callback) {
-	var param = {text: text, user: user}
-	MessageModel.update(param, {
+messages.report = function(messageID, callback) {
+	console.log(messageID);
+	MessageModel.update({id: messageID}, {
 		reported: true
-	}, function(){
-		callback();
-	});
+	}, callback);
 }
 
-messages.free = function(user, text, callback) {
-	var param = {text: text}; //, user: user}
-	MessageModel.update(param, {
+messages.free = function(messageID, callback) {
+	MessageModel.update({id: messageID}, {
 		reported: false
-	}, function(){
-		callback();
-	});
+	}, callback);
 }
 
-messages.delete = function(user, text, callback) {
+messages.delete = function(messageID, callback) {
 	//all distasteful messages regardless of user should be deleted
-	var param = {text: text};//, user: user};
-	MessageModel.remove(param, function(){
+	MessageModel.remove({id: messageID}, function(){
 		callback();
 	});
 }
 
-messages.next = function(id, volume, callback) {
-	var param = id != "" ? {_id:{"$lt": mongoose.Types.ObjectId(id)}} : null;
+messages.next = function(messageID, volume, callback) {
+	var param = messageID ? {id:{"$lt": messageID}} : null;
 	var promise = MessageModel.find(param).limit(parseInt(volume)).sort({"_id":-1}); 
 	promise.then(function(data){
 		callback(data);
 	})
 }
 
+messages.generateUsername = function() {
```

Comment:
The reason the way I did it the way I did it was to address ^.  The probability of getting a username already assigned increases as more usernames are distributed.

Let's say 50,000 unique visits occur within the span of a year.  That means, on average, the db (which is usually the performance bottleneck) will have to retrieve 2 times. 

Let's say after another year, we're at 90,000 unique visits, we will on average have to query the db 10 times to get a free username.  I know these seem like large numbers, but the reason my logic wasn't reduced to this line in the first place was to ensure that something like this wouldn't happen, at the cost of longer usernames.  Each increase of 3 more digits implies x1000 more available unique values, and a probability of obtaining a unique username already assigned reduced by x1000.

Let me know if this doesn't make sense.

Again, I know these are quite large numbers, but socket.io will distribute a username upon each unique visitor (unique laptop, phone, tablet, etc).  I want it this way so nothing deters the user into participating.

An alternative solution I thought of was having usernames expire on their own in the DB.